msprime/species_tree.py

```diff
@@ -0,0 +1,834 @@
+    resulting instances of PopulationConfiguration.
+
+    Introgression events, as stored in the extended Newick format by
+    PhyloNet and possibly other programs, are not parsed and therefore
+    not used.
+    Rate matrices of continuous migration between co-existing lineages
+    can not be stored in (extended) Newick or Nexus format and are
+    therefore also not used here. However, these can be added to the
+    model after population configurations and demographic events are
+    defined with this function.
+    """
+
+    # Make sure a filename is specified.
+    err = 'A filename must be specified when calling function parse_species_tree().'
+    assert filename, err
+
+    # Make sure that branch length units are either "myr", "yr", or "gen".
+    allowed_branch_lenth_units = ["myr", "yr", "gen"]
+    if branch_length_units not in allowed_branch_lenth_units:
+        err = 'The specified units for branch lengths ('
+        err += '"{}") are not accepted. '.format(branch_length_units)
+        err += 'Accepted units are "myr" (millions of years), "yr" (years), '
+        err += 'and "gen" (generations).'
+        raise ValueError(err)
+
+    # Make sure that the sample size is either None or non-negative.
+    if sample_size is not None and sample_size < 0:
+        raise ValueError("Sample size must be >= 0")
+
+    # Make sure that the population size is either None or positive.
+    if Ne is not None and Ne <= 0:
+        raise ValueError("Population size Ne must be > 0")
+
+    # Make sure that the generation time is either None or positive.
+    if generation_time is not None and generation_time <= 0:
+        raise ValueError("Generation time must be > 0")
+
+    # Make sure that the generation time is specified if and only if
+    # branch lengths are not in units of generations.
+    if branch_length_units == "gen":
+        err = 'With branch lengths in units of generations ("gen"), generation_time '
+        err += 'should not be specified.'
+        assert generation_time is None, err
+    else:
+        err = 'With branch lengths in units of '
+        err += '"{}", generation_time must be specified.'.format(branch_length_units)
+        assert generation_time, err
+
+    # Get the number of generations per branch length unit.
+    if branch_length_units == "gen":
+        generations_per_branch_length_unit = 1
+    elif branch_length_units == "myr":
+        generations_per_branch_length_unit = 1000000/generation_time
+    else:
+        generations_per_branch_length_unit = 1/generation_time
+
+    # Read the input file.
+    tree_string = None
+    starbeast_format = False
+    starbeast_root_pop_size = None
+    infile = open(filename)
+    inlines = infile.readlines()
+    assert len(inlines) > 0, 'File {} is empty.'.format(filename)
+    if inlines[0][0:6].lower() == '#nexus':
+        # Assume the input is in nexus format. Maximally one tree string is read.
```

Comment:
Let's split this out into a function, ``parse_nexus``

Reply:
The new functions `parse_species_tree` and `parse_starbeast` now allow input that is either a one-line Newick string or a Nexus file written by StarBEAST. Nexus files that are not in StarBEAST format are not allowed. If we stick to this, the change that @jeromekelleher suggests here is no longer relevant.